patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch

```diff
@@ -306,7 +306,37 @@
        if (p_217401_2_ != null) {
           explosion.func_199592_a(p_217401_2_);
        }
-@@ -1377,4 +1421,15 @@
+@@ -1109,12 +1153,14 @@
+    }
+ 
+    public <T extends IParticleData> int func_195598_a(T p_195598_1_, double p_195598_2_, double p_195598_4_, double p_195598_6_, int p_195598_8_, double p_195598_9_, double p_195598_11_, double p_195598_13_, double p_195598_15_) {
+-      SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, (float)p_195598_9_, (float)p_195598_11_, (float)p_195598_13_, (float)p_195598_15_, p_195598_8_);
++      net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(p_195598_1_, false, p_195598_8_, p_195598_2_, p_195598_4_, p_195598_6_, p_195598_9_, p_195598_11_, p_195598_13_, p_195598_15_);
++      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return 0;
++      SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
+       int i = 0;
+ 
+       for(int j = 0; j < this.field_217491_A.size(); ++j) {
+          ServerPlayerEntity serverplayerentity = this.field_217491_A.get(j);
+-         if (this.func_195601_a(serverplayerentity, false, p_195598_2_, p_195598_4_, p_195598_6_, sspawnparticlepacket)) {
++         if (this.func_195601_a(serverplayerentity, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket)) {
+             ++i;
+          }
+       }
+@@ -1123,8 +1169,10 @@
+    }
+ 
+    public <T extends IParticleData> boolean func_195600_a(ServerPlayerEntity p_195600_1_, T p_195600_2_, boolean p_195600_3_, double p_195600_4_, double p_195600_6_, double p_195600_8_, int p_195600_10_, double p_195600_11_, double p_195600_13_, double p_195600_15_, double p_195600_17_) {
+-      IPacket<?> ipacket = new SSpawnParticlePacket(p_195600_2_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, (float)p_195600_11_, (float)p_195600_13_, (float)p_195600_15_, (float)p_195600_17_, p_195600_10_);
+-      return this.func_195601_a(p_195600_1_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, ipacket);
++      net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(p_195600_2_, false, p_195600_10_, p_195600_4_, p_195600_6_, p_195600_8_, p_195600_11_, p_195600_13_, p_195600_15_, p_195600_17_);
```

Comment:
Again this can probably be handled better by using a method in `ForgeEventFactory` possibly returning an Optional which is empty when the event is cancelled, or contains the even if it is allowed to pass.

Would again reduce the patch size.

Reply:
I'm having a bit of trouble understanding how the optional will reduce patch size. Using an optional I get this in the ServerWorld.java:
```
public <T extends IParticleData> int spawnParticle(T type, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
      AtomicInteger i = new AtomicInteger();
      ForgeEventFactory.onWorldParticleSpawn(type, false, posX, posY, posZ, particleCount, xOffset, yOffset, zOffset, speed).ifPresent(event -> {
         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
         for (ServerPlayerEntity serverplayerentity : this.players) {
            if (this.sendPacketWithinDistance(serverplayerentity, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket)) {
               i.incrementAndGet();
            }
         }
      });
      return i.get();
   }

   public <T extends IParticleData> boolean spawnParticle(ServerPlayerEntity player, T type, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
      AtomicBoolean result = new AtomicBoolean(false);
      ForgeEventFactory.onWorldParticleSpawn(type, longDistance, posX, posY, posZ, particleCount, xOffset, yOffset, zOffset, speed).ifPresent(event -> {
         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
         result.set(this.sendPacketWithinDistance(player, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket));
      });
      return result.get();
   }
```

With the following in the ForgeEventFactory.java:
```
public static Optional<ParticleWorldHandleEvent> onWorldParticleSpawn(IParticleData particle, boolean longDistance, double posX, double posY, double posZ, int particleCount, double xOffset, double yOffset, double zOffset, double speed) {
        net.minecraftforge.event.ParticleWorldHandleEvent event = new net.minecraftforge.event.ParticleWorldHandleEvent(particle, longDistance, particleCount, speed, posX, posY, posZ, xOffset, yOffset, zOffset);
        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return Optional.empty();
        return Optional.of(event);
    }
```